daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala

```diff
@@ -202,6 +212,29 @@ final case class DefTemplate[+Ty](
   }
 
   def getKey: j.Optional[_ <: Ty] = toOptional(key)
+
+  private[iface] def extendWithInterface[OTy >: Ty](
+      ifaceName: Ref.TypeConName,
+      ifc: DefInterface[OTy],
+  ): DefTemplate[OTy] = {
+    import TemplateChoices.{Resolved, Unresolved}
+    copy(
+      implementedInterfaces = implementedInterfaces :+ ifaceName,
+      tChoices = tChoices match {
+        case unr @ Unresolved(_, sources) =>
+          unr.copy(unresolvedChoiceSources = sources incl ifaceName)
+        // If unresolved, we need only add ifc as a future interface to resolve;
+        // otherwise, we must self-resolve and add to preexisting resolutions
+        case r @ Resolved(rc) =>
+          type K[C] = Semigroup[Resolved.Choices[C]]
```

Comment:
Why is this needed?

Reply:
We have a sensible semigroup for almost the whole resolved choices structure, but we need a way to "combine" single choices; the rule we choose is "just use the first one". We use `K` to specify that `FirstVal` is our chosen rule.